component/namespace-quota.jsonnet

```
@@ -0,0 +1,117 @@
+  deny: {
+    conditions: [
+      {
+        key: '{{nsCount}}',
+        operator: 'GreaterThanOrEquals',
+        value: '{{override || `%s`}}' % params.maxNamespaceQuota,
+      },
+    ],
+  },
+};
+
+local orgLabelVar = '{{request.object.metadata.labels."appuio.io/organization"}}';
+
+/**
+  * Check Namespace Quota
+  * This policy will:
+  * - Deny the new namespace if the number of existing namespaces is greater or equal a certain number.
+  * - This number is either a default defined in this component, or it can be overridden for a specific organization.
+  *   To set this override, create a config map in the component namespace with name pattern 'override-<org-name>' with `.data.namespaceOverride` being the number.
+  *   For example: kubectl -n appuio-cloud create cm override-foo --from-literal=namespaceQuota=4
+  */
+local namespaceQuotaPolicy = kyverno.ClusterPolicy('check-namespace-quota') {
+  spec: {
+    validationFailureAction: 'enforce',
+    background: false,
+    rules: [
+      {
+        name: 'check-namespace-count',
+        match: common.MatchOrgNamespaces,
+        exclude: common.BypassNamespaceRestrictionsSubjects(),
+        context: [
+          overrideContext(varName=orgLabelVar),
+          nsCountContext(varName=orgLabelVar),
+        ],
+        validate: validateRule(varName=orgLabelVar),
+      },
+      {
+        name: 'check-project-count',
+        match: common.MatchProjectRequests(),
+        exclude: common.BypassNamespaceRestrictionsSubjects(),
+        context: [
+          {
+            name: 'organization',
+            apiCall: {
+              urlPath: '/apis/user.openshift.io/v1/users/{{request.userInfo.username}}',
+              jmesPath: 'metadata.annotations."appuio.io/default-organization"',
+            },
+          },
+          overrideContext(varName='{{organization}}'),
+          nsCountContext(varName='{{organization}}'),
+        ],
+        validate: validateRule(varName='{{organization}}'),
+      },
+    ],
+  },
+  metadata+: {
+    annotations+: {
+      // Workaround for bug in `kyverno-cli test` command.
+      // Commodore generates an empty annotation object (`annotations: {}`) which trips up Kyverno.
+      // A non empty object or `annotations: null` is valid.
+      // Ensure a label to not have an empty object.
+      'policies.kyverno.io/category': 'Namespace Management',
+    },
+  },
+};
+
+local namespaceQuotaOverrides = [
+  kube.ConfigMap('override-%s' % org) {
+    metadata+: {
+      labels+: {
+        // this label is informational only here
+        'appuio.io/organization': org,
+      },
+    },
+    data+: {
+      namespaceQuota: '%s' % params.namespaceQuotaOverrides[org],
+    },
+  } + common.DefaultLabels
+  for org in std.filter(function(key) key != null && params.namespaceQuotaOverrides[key] != null, std.objectFields(params.namespaceQuotaOverrides))
```

There's also `std.prune()`:

> Recursively remove all "empty" members of a. "Empty" is defined as zero length `arrays`, zero length `objects`, or `null` values. The argument a may have any type.

```suggestion
  for org in std.objectFields(std.prune(params.namespaceQuotaOverrides))
```

---

Yes, but our best practice discourages it: https://syn.tools/syn/explanations/jsonnet.html#_pruning_lists_and_objects

---

As far as I understand it this best practice is for `filter+prune`.

I think prune is way easier to read and therefore I prefer it. No super strong feelings.

As a sidenote: The performance differences _only_ apply if it's a deep nested object like filtering Kubernetes manifests from the root.

---

I agree it's not really necessary since we don't expect large nested objects here.
However, at the same time I don't want to set a bad example of using `prune` when our own guidelines discourage it. This code snippet might be copy-pasted without thinking about it.
And also for consistency reasons I'd rather avoid it.

---

I don't agree to optimise for a use-case that most likely will never happen: Someone copies code from exactly this place and puts it somewhere with a strange `filter+prune` construct. 

We already use `std.prune` all over the place: https://github.com/search?q=org%3Aprojectsyn+std.prune&type=code

But it's a minor issue and personal preference so i'll resolve this.

---

> We already use std.prune all over the place

Yes, because we didn't know better at that time, they're leftover. Many components in that list haven't really been touched since a long time.

It's less about the copy-pasting. I just feel our guidelines become worth nothing if we don't follow it. Though maybe the guidelines need updating to cover such a case this here?